### PR TITLE
[MRG] add max_containment to `MinHash` class.

### LINKED ISSUE #1247: add name, md5 of query signature to gather and search output
courtesy of Courtney's rotation project, sometimes you want to do big database searches and combine all CSVs and figure out which ones match to which query, and right now you can't...

### LINKED ISSUE #1343: consider adding --max-containment option for some sourmash utilities
Given the work we've been doing with maximum containment --> genome distance metrics, there are some places where it might be useful to return the maximum containment between signatures, rather than directional containment. This could be enabled with a `--max-containment` option for certain commands.

Cases I can think of:

- `compare`
- `search` -- e.g. my current desired use case is `sourmash search --max-containment --best-only` to find the best match for an input sig (or list of input sigs) to a list of cluster founders.

`--max-containment` is likely _not_ useful for gather-style metagenome applications, primarily bc the direction of containment (intersection/reference genome hashes) is already ideal.

### PR DESCRIPTION
**Pull request by ctb**
Add `max_containment()` and `--max-containment` per #1343.

Fixes https://github.com/dib-lab/sourmash/issues/1247
Fixes #1343

Main features:
* adds `MinHash.max_containment(other, downsample=False)`
* adds `SourmashSignature.max_containment(other, downsample=False)`
* adds `--max-containment` flag to `sourmash search`
* adds `do_max_containment` flag to search for `Index` classes (`LinearIndex`, `SBT`, `LCA_Database`)

In addition, this PR indulges in misc cleanup:
* refactors some of the `Index` code to use named arguments, now that legacy support for Python 2.7 was removed.
* removes untested `result` caching in `sbtmh.py`.
* fixes namespace collision and associated broken test in `test_signature.py` - there were two `test_str` functions.
* fixes namespace collision and associated broken test in `test__minhash.py` -- there were two `test_mh_len` functions.
* test that `SBT.search(...)` properly fails when `threshold=None` (this code did not have test coverage)
* remove duplicate `test_compare_containment_abund_flatten` test function in `test_sourmash.py`.

TODO:
- [x] add feature for `compare --max-containment` along with tests
- [x] check that `max_containment` does not allow non-scaled signatures
- [x] add test for `if do_containment and do_max_containment` conditions
- [x] add test for `if threshold is None:` in sbt.py
- [x] add command line tests for max containment
- [ ] check SBT especially

## Checklist

- [ ] Is it mergeable?
- [ ] `make test` Did it pass the tests?
- [ ] `make coverage` Is the new code covered?
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

### CODE REVIEW

File: src/sourmash/index.py [hunk 1]
```diff
@@ -60,11 +60,14 @@ def search(self, query, *args, **kwargs):
...
 
         do_containment = kwargs.get('do_containment', False)
+        do_max_containment = kwargs.get('do_max_containment', False)
         ignore_abundance = kwargs.get('ignore_abundance', False)
 
         # configure search - containment? ignore abundance?
         if do_containment:
             query_match = lambda x: query.contained_by(x, downsample=True)
+        elif do_max_containment:
+            query_match = lambda x: query.max_containmenty(x, downsample=True)
```
Comment (bluegenes) [inline src/sourmash/index.py]:
```suggestion
            query_match = lambda x: query.max_containment(x, downsample=True)
```

**Comment by ctb** [comment]
# witness!

## search
```
% sourmash search genome-s10.fa.gz.sig all.sbt.zip  

== This is sourmash version 4.0.0rc2.dev1+gce950caa. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

select query k=31 automatically.
loaded query: ../genome-s10.fa.gz... (k=31, DNA)
loaded 1 databases.

3 matches:
similarity   match
----------   -----
100.0%       ../genome-s10.fa.gz
 51.6%       ../genome-s10+s11.fa.gz
 16.7%       ../genome-s10-small.fa.gz
```

## search --containment

```
sourmash search genome-s10.fa.gz.sig all.sbt.zip  --containment

== This is sourmash version 4.0.0rc2.dev1+gce950caa. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

select query k=31 automatically.
loaded query: ../genome-s10.fa.gz... (k=31, DNA)
loaded 1 databases.

3 matches:
similarity   match
----------   -----
100.0%       ../genome-s10.fa.gz
100.0%       ../genome-s10+s11.fa.gz
 16.7%       ../genome-s10-small.fa.gz
```

## search --max-containment

```
% sourmash search genome-s10.fa.gz.sig all.sbt.zip  --max-containment

== This is sourmash version 4.0.0rc2.dev1+gce950caa. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

select query k=31 automatically.
loaded query: ../genome-s10.fa.gz... (k=31, DNA)
loaded 1 databases.

3 matches:
similarity   match
----------   -----
100.0%       ../genome-s10-small.fa.gz
100.0%       ../genome-s10.fa.gz
100.0%       ../genome-s10+s11.fa.gz
```

**Comment by ctb** [comment]
@bluegenes this might be ready to try out. I wouldn't trust the SBT code just yet, but everything else should work I think.

**Comment by bluegenes** [comment]
Thanks @ctb!! Will take it for a spin

**Comment by bluegenes** [comment]
`sourmash search --max-containment` seems to be running on my testing! YAY.

I somewhat forgot that the `search` csv output is less informative than the `gather` csv output -- ultimately, it would be really nice to have add'l info, particularly signature name + matched bp, etc, as you suggest in #1366 for 5.0. 

 
some mini results:

` sourmash search --max-containment output.protein-pigeon/prodigal/signatures/pigeon1.0-GOV_18503.prodigal.sig output.pigeon1.0-cluster/pigeon1.0.protein-k10.mc0.05.founders.sbt.zip -o output.pigeon1.0-cluster/cluster_info/GOV_18503_x_pigeon1.0.protein-k10.mc0.05.best-founder.txt --threshold 0.01 --ksize 10 --protein`

`--max-containment`:
```
4 matches; showing first 3:
similarity   match
----------   -----
 10.0%       EarthsVirome_13607
  8.7%       GOV_66421
  4.5%       EarthsVirome_15562
```
** note, none of these show up with the default threshold (`0.08`). I needed `--threshold 0.05` for the first match to show up, `0.03` for the next and `0.02` for the 3rd match. I guess I thought `search` used a percentage threshold -- will go look at it, but wanted to record here in the mean time.

`--containment`:
```
4 matches; showing first 3:
similarity   match
----------   -----
  5.3%       EarthsVirome_13607
  3.1%       GOV_66421
  2.3%       EarthsVirome_15562
```

jaccard:
```
3 matches:
similarity   match
----------   -----
  3.8%       EarthsVirome_13607
  2.7%       GOV_66421
  1.7%       EarthsVirome_15562
```

**Comment by bluegenes** [comment]
would it be straightforward to add `compare --max-containment` here, or would that be better in a separate PR?

**Comment by ctb** [comment]
> would it be straightforward to add `compare --max-containment` here, or would that be better in a separate PR?

Totes. Great idea.

**Comment by ctb** [comment]
> `sourmash search --max-containment` seems to be running on my testing! YAY.
> 
> I somewhat forgot that the `search` csv output is less informative than the `gather` csv output -- ultimately, it would be really nice to have add'l info, particularly signature name + matched bp, etc, as you suggest in #1366 for 5.0.

Yeah, `search` CSV output is *terrible*! As you will see in https://github.com/dib-lab/sourmash/pull/1370 😉 I am planning to output the following in the `prefetch` command,
* intersect bp
* match bp
* query bp
* jaccard similarity
* query containment
* match containment
* max_containment
and the only hesitation I have in adding all of these to `search` output is that they would be subject to different thresholds depending on the command line arguments, and my hot take is that I don't like that.

Specifically, imagine that you do a search with `--max-containment` vs `--similarity` - for the same query, the results will (of course) be different, and the order and thresholding in the output CSV will be different.

Of course, this is a problem for the current output, too, in that the actual meaning of the `similarity` number in the output file changes depending on the search parameters. So that's bad. Grr.

🤔 

Hmm, I wonder if we could include an extra column that is something like "search key" to indicate what the search was? It'll be redundant (since it would have to be in every row for this columnar data output) but at least it'd be there.

Might also be a good opportunity to support JSON or YAML output from search, gather, and prefetch - then those files could contain far more information, including full command-line parameters, threshold, etc. etc. See #448.

We could also turn off CSV output from search entirely, and tell people to use prefetch for programmatic foo...

> ** note, none of these show up with the default threshold (`0.08`). I needed `--threshold 0.05` for the first match to show up, `0.03` for the next and `0.02` for the 3rd match. I guess I thought `search` used a percentage threshold -- will go look at it, but wanted to record here in the mean time.

Curious what you mean by "percentage threshold"? It just takes a floating point number that is the lowest similarity etc to report.

OH! You mean the thresholding is actually just wrong for `--max-containment`. I, umm, should fix that! I appreciate the implication that this was actually something intelligent, and not just a dumb bug...

**Comment by ctb** [comment]
hi @bluegenes this PR is ready for review (and merge?)!

Other than code review (sorry for all the mess...) the two remaining things are --

* do you want me to add more stuff to the CSV output in this PR, or should I start a new PR for that? I think this one is getting kind of big.
* could do a test on a large-ish SBT, by doing `sourmash search --max-containment` on both the SBT and on a directory of contents created with `sourmash sig split` run on the SBT? I'd like a double-check on the SBT max containment code - no reason it shouldn't work, just slightly paranoid about computers...

**Comment by ctb** [comment]
I added the query information into the search CSV output in cbd25037. I couldn't add all of the info that @bluegenes might want in, however, because `search` works on regular MinHashes, too, which don't support containment or max_containment.

Perhaps another reason to consider getting rid of "regular" MinHash per https://github.com/dib-lab/sourmash/issues/1354

**Comment by ctb** [comment]
(I suppose I could put zeros in for all those numbers when running on regular minhashes...)

**Comment by bluegenes** [comment]
>(I suppose I could put zeros in for all those numbers when running on regular minhashes...)

NA's?

**Comment by ctb** [comment]
On Thu, Mar 11, 2021 at 05:01:40PM -0800, Tessa Pierce wrote:
> >(I suppose I could put zeros in for all those numbers when running on regular minhashes...)
> 
> NA's?

<sigh>

File: src/sourmash/commands.py [hunk 2]
```diff
@@ -91,16 +92,24 @@ def compare(args):
...
+        error('must use scaled signatures with --containment and --max-containment')
         sys.exit(-1)
 
     # notify about implicit --ignore-abundance:
-    if args.containment:
+    if is_containment:
         track_abundances = any(( s.minhash.track_abundance for s in siglist ))
         if track_abundances:
-            notify('NOTE: --containment means signature abundances are flattened.')
+            notify('NOTE: --containment and --max-containment means signature abundances are ignored.')
```
Comment (bluegenes) [inline src/sourmash/commands.py]:
```suggestion
            notify('NOTE: --containment and --max-containment ignore signature abundances.')
```

File: src/sourmash/commands.py [hunk 3]
```diff
@@ -477,7 +497,8 @@ def search(args):
         notify("** reporting only one match because --best-only was set")
 
     if args.output:
-        fieldnames = ['similarity', 'name', 'filename', 'md5']
+        fieldnames = ['similarity', 'name', 'filename', 'md5',
+                      'query_filename', 'query_name', 'query_md5']
```
Comment (bluegenes) [inline src/sourmash/commands.py:501]: It would be nice to modify `similarity` to `containment` / `max_containment` for csv output. Thoughts?

Comment (ctb) [reply src/sourmash/commands.py:501]: semantic versioning prevents us from removing the `similarity` header before 5.0. we could add new columns, I 'spose. I don't like the idea that column headers **change** depending on command line arguments, though. Not sure how to think about it.

Comment (ctb) [reply src/sourmash/commands.py:501]: (suggest we punt this to a new issue and discuss it there.)

File: src/sourmash/minhash.py [hunk 4]
```diff
@@ -543,6 +543,18 @@ def contained_by(self, other, downsample=False):
 
         return self.count_common(other, downsample) / len(self)
 
+    def max_containment(self, other, downsample=False):
+        """\
```
Comment (bluegenes) [inline src/sourmash/minhash.py]:
```suggestion
        """
```

File: src/sourmash/minhash.py [hunk 5]
```diff
@@ -543,6 +543,18 @@ def contained_by(self, other, downsample=False):
...
+        Calculate maximum containment.
+        """
+        if not (self.scaled and other.scaled):
+            raise TypeError("can only calculate containment for scaled MinHashes")
+        min_denom = min((len(self), len(other)))
+        if not min_denom:
+            return 0.0
```
Comment (bluegenes) [inline src/sourmash/minhash.py:554]: this should really never happen, right? Or could an empty sketch or aggressive downsampling trigger this?

File: src/sourmash/sbt.py [hunk 6]
```diff
@@ -378,13 +383,21 @@ def search(self, query, *args, **kwargs):
         if do_containment:
             search_fn = search_minhashes_containment
             query_match = lambda x: tree_query.contained_by(x, downsample=True)
+        elif do_max_containment:
+            search_fn = search_minhashes_max_containment
+            query_match = lambda x: tree_query.max_containment(x,
+                                                               downsample=True)
 
         if best_only:            # this needs to be reset for each SBT
             search_fn = SearchMinHashesFindBest().search
```
Comment (bluegenes) [inline src/sourmash/sbt.py:394]: does this need to change at all for containment?

I see jaccard over here? https://github.com/dib-lab/sourmash/blob/ca201cfc1824900bfb94aaadd94e163c70b7cf4e/src/sourmash/sbtmh.py#L135

File: src/sourmash/minhash.py (hunk 5 above)
Comment (ctb) [reply src/sourmash/minhash.py:554]: exactly - empty sketch, perhaps triggered by aggressive downsampling.

In 18b4f780, I've added two tests, `test_search_containment_s10_sbt_empty` and `test_search_max_containment_s10_sbt_empty `, that trigger this condition. Thanks!

Comment (bluegenes) [reply src/sourmash/minhash.py:554]: thx!

File: src/sourmash/search.py [hunk 7]
```diff
@@ -25,33 +25,32 @@ def format_bp(bp):
...
+        x.append(SearchResult(similarity=score,
                               match=match,
                               md5=match.md5sum(),
                               filename=filename,
-                              name=match.name))
+                              name=match.name,
+                              query=query,
+                              query_filename=query.filename,
+                              query_name=query.name,
+                              query_md5=query.md5sum()[:8]
```
Comment (bluegenes) [inline src/sourmash/search.py:52]: is there a reason to truncate one md5 and not the other?

...and now that I think about it, it's for semantic versioning, eh?

File: src/sourmash/sbt.py (hunk 6 above)
Comment (ctb) [reply src/sourmash/sbt.py:394]: good catch, this is straight up a bug - when `--best-only` is specified, it reverts to calculating Jaccard similarity on SBTs!

I added an exception raise and some tests in 108748a4 to prevent this for now, will punt this to a new issue to fix properly.

File: src/sourmash/search.py (hunk 7 above)
Comment (ctb) [reply src/sourmash/search.py:52]: err, nope, didn't notice I _wasn't_ truncating the md5sum for the match. Hrm.

**Review by bluegenes — APPROVED** [review]
the rest lgtm!

File: src/sourmash/commands.py (hunk 3 above)
Comment (ctb) [reply src/sourmash/commands.py:501]: punted to https://github.com/dib-lab/sourmash/issues/1390

File: src/sourmash/search.py (hunk 7 above)
Comment (ctb) [reply src/sourmash/search.py:52]: punted to https://github.com/dib-lab/sourmash/issues/1390

**Comment by ctb** [comment]
OK, I _think_ I addressed everything. Will wait for tests to pass before merging.
